hoss/dimension_utilities.py

```diff
@@ -75,8 +75,150 @@ def prefetch_dimension_variables(opendap_url: str, varinfo: VarInfoFromDmr,
+    add_bounds_variables(required_dimensions_nc4, required_dimensions,
+                         varinfo, logger)
+
+    return required_dimensions_nc4
+
+
+def add_bounds_variables(dimensions_nc4: str,
+                         required_dimensions: Set[str],
+                         varinfo: VarInfoFromDmr,
+                         logger: Logger) -> None:
+    """ Augment a NetCDF4 file with artificial bounds variables for each 
+        dimension variable that is edge-aligned and does not already 
+        have bounds variables. 
+
+        For each dimension variable:
+        (1) Check if the variable needs a bounds variable.
+        (2) If so, create a bounds array from within the `write_bounds` function.
+        (3) Then write the bounds variable to the NetCDF4 URL.
+
+    """
+    with Dataset(dimensions_nc4, 'r+') as prefetch_dataset:
+        for dimension_name in required_dimensions:
+            dimension_variable = varinfo.get_variable(dimension_name)
+            if needs_bounds(dimension_variable) is True:
+                write_bounds(prefetch_dataset, dimension_variable)
+
+                logger.info('Artificial bounds added for dimension variable: '
+                f'{dimension_name}')
+
+
+def needs_bounds(dimension: VariableFromDmr) -> bool:
+    """ Check if a dimension variable needs a bounds variable.
+        This will be the case when dimension cells are edge-aligned 
+        and bounds for that dimension do not already exist.
+    
+    """
+    return (
+        dimension.attributes.get('cell_alignment') == 'edge'
+        and dimension.references.get('bounds') is None
+    )
+
+
+def get_bounds_array(prefetch_dataset: Dataset,
+                     dimension_path: str) -> np.ndarray:
+    """ Create an array containing the minimum and maximum bounds
+        for each pixel in a given dimension. 
+
+        The minimum and maximum values are determined under the assumption
+        that the dimension data is monotonically increasing and contiguous.
+        So for every bounds but the last, the bounds are simply extracted
+        from the dimension dataset.
+
+        The final bounds must be calculated with the assumption that 
+        the last data cell is edge-aligned and thus has a value the does
+        not account for the cell length. So, the final bound is determined
+        by taking the median of all the resolutions in the dataset to obtain
+        a resolution that can be added to the final data value. 
+
+        Ex: Input dataset with resolution of 3 degrees:  [ ... , 81, 84, 87]
+
+        Minimum | Maximum
+         <...>     <...>
+          81        84
+          84        87
+          87        ?  ->  87 + median resolution -> 87 + 3 -> 90
+        
+    """
+    # Access the dimension variable's data using the variable's full path.
+    dimension_array = prefetch_dataset[dimension_path][:]
+
+    median_resolution = np.median(np.diff(dimension_array))
```

Comment:
> Is it always the case that when something is edge aligned, the values are such that they represent the left side of the gridcell? 
No, I'm pretty sure any type of edge alignment is possible. We decided to stick with handling the lower left alignment case since that;s ATL16's alignment and accounting for all the possible alignments could balloon very quickly. _I think_ we're hoping that this is the most common edge-alignment. I might have just made some of that up though, what say you @owenlittlejohns?

Reply:
>Doesn't the dimension array have constant differences between cells?

@autydp and @owenlittlejohns looked over this calculation with me, and they figured the cell differences might not always be constant. So yes, what I currently have here is a "good enough". They can weigh in though!

Reply:
I might still argue that the (N-2, N-1) difference is a better predictor of the (N-1, N) difference than the entire median, but I will not push it.

Reply:
I think the main thing here is that there can be a bit of floating-point level variation. Maybe that's not _super_ important for the very last pixel, but it's bitten us before at times for requests with bounding boxes hitting the exact boundary between pixels. My preference for the median diff comes from a combination of that, and that it's still all a pretty straightforward one-liner to get it with `numpy`.